modules/fedimint-wallet-client/src/backup.rs

```diff
@@ -411,11 +412,18 @@ where
                 .advance(RECOVER_NUM_IDX_ADD_TO_LAST_USED);
         }
 
-        let history = check_addr_history(cur_tweak_idx).await?;
-
-        if !history.is_empty() {
+        if used_tweak_idxes.contains(&cur_tweak_idx) {
+            debug!(target: LOG_CLIENT_MODULE_WALLET,
+                tweak_idx=%cur_tweak_idx,
+                "Skipping checking history of an address, as it was previously used");
             last_used_idx = Some(cur_tweak_idx);
-            tweak_idxes_with_pegins.insert(cur_tweak_idx);
+        } else {
+            let history = check_addr_history(cur_tweak_idx).await?;
+
+            if !history.is_empty() {
+                last_used_idx = Some(cur_tweak_idx);
+                tweak_idxes_with_pegins.insert(cur_tweak_idx);
+            }
```

Comment:
The amount of mutability here makes it quite hard to review, I was procrastinating on reviewing this PR for way too long due to it. Let's discuss during review club, I think there is a beautiful, small algorithm hidden in here that wants to get implemented with some iterator chaining :laughing:

Reply:
call: could loop over a range

Reply:
@elsirion Hmm... now that I'm looking at it ... if I do `while` or `for`, I can't use `break value;` which will introduce a different mutable value.